cpp/include/cudf/detail/utilities/vector_factories.hpp

```diff
@@ -28,6 +30,51 @@
 namespace cudf {
 namespace detail {
 
+/**
+ * @brief Asynchronously construct a `device_uvector` and set all elements to zero.
+ *
+ * @note This function does not synchronize `stream`.
+ *
+ * @tparam T The type of the data to copy
+ * @param size The number of elements in the created vector
+ * @param stream The stream on which to allocate memory and perform the memset
+ * @param mr The memory resource to use for allocating the returned device_uvector
+ * @return A device_uvector containing zeros
+ */
+template <typename T>
+rmm::device_uvector<T> make_zeroed_device_uvector_async(
```

Comment:
Mark, should we have a generic version of `make_device_uvector_xxx` taking a parameter `T const& default_value = T{0}` instead? Here we only have zero but can't initialize the default value with something else (which we really need).

Reply:
It's impossible to make such an API asynchronous as it requires copying from a temporary which may go out of scope before the copy completes.

Reply:
We can provide an async fill, or contribute one to thrust. Out of scope of this PR though.